Enabling console logger queue length and mode to be configurable

maryamariyan opened the pull request:
### Description

Adds configuration for console logger's queue length and mode

- Adds new APIs for the logging console queue processor
- Replaces BlockingCollection with Queue implementation
- Adds code coverage for ConsoleLoggerProcessor (except for for `ThreadStateException` case in Dispose and `GetQueueSize` metrics logic)
- Adds queue name and metrics

### Performance testing

#### Comparing original `BlockingCollection` with new `Queue` implementation

Using a custom log formatter which only writes the string in message template, with no padding or prefix, using the snippet below:

```c#
        var howManyBytes = _1000Chars.Length * sizeof(char);
        var logger = loggerFactory.CreateLogger<Program>();
        var sw = new Stopwatch();
        sw.Start();
        for (int i = 0; i < 100000; i++)
        {
            logger.LogInformation(_1000Chars);
        }
        sw.Stop();
        var elapsedForLoggingConsole = sw.ElapsedMilliseconds;

        Console.Error.WriteLine($"Bytes written {howManyBytes * 100000 / (1024 * 1024)} MB");
        Console.Error.WriteLine($"elapsedForLoggingConsole: {elapsedForLoggingConsole} ms");
        Console.Error.WriteLine($"Speed: {(howManyBytes * 100000 / (1024 * 1024)) / (elapsedForLoggingConsole / 1000.0f)} MB/sec");
```

Table below shows the perf numbers comparing `BlockingCollection` vs. new `Queue` implementation given different max length sizes, running:

```
> dotnet run -c Release >> output-to-file.txt
```

```
|           Queue type | Max length |  Time elapsed |            Speed |   Bytes written |      Notes        |
|----------------------|------------|---------------|------------------|-----------------|-------------------|
|   BlockingCollection |         1  |      4319 ms  |  43.99167 MB/sec |          190 MB | Fastest (10 runs) |
|   BlockingCollection |         1  |      4546 ms  | 41.794983 MB/sec |          190 MB | Slowest (10 runs) |
|                Queue |         1  |      4309 ms  | 44.093758 MB/sec |          190 MB | Fastest (10 runs) |
|                Queue |         1  |      5078 ms  | 37.416306 MB/sec |          190 MB | Slowest (10 runs) |
|   BlockingCollection |        10  |      4269 ms  |  44.50691 MB/sec |          190 MB | Fastest (10 runs) |
|   BlockingCollection |        10  |      4482 ms  | 42.391792 MB/sec |          190 MB | Slowest (10 runs) |
|                Queue |        10  |      4374 ms  |   43.4385 MB/sec |          190 MB | Fastest (10 runs) |
|                Queue |        10  |      4496 ms  |  42.25979 MB/sec |          190 MB | Slowest (10 runs) |
|   BlockingCollection |      1024  |      4295 ms  |  44.50691 MB/sec |          190 MB | Fastest (10 runs) |
|   BlockingCollection |      1024  |      4574 ms  | 41.539135 MB/sec |          190 MB | Slowest (10 runs) |
|                Queue |      1024  |      4223 ms  |  44.99171 MB/sec |          190 MB | Fastest (10 runs) |
|                Queue |      1024  |      4675 ms  |  40.64171 MB/sec |          190 MB | Slowest (10 runs) |
```

This shows there's no time regression caused by using the queue approach. All of the cases above would hit the max queue length limit and they all have a perf hit because of hitting that limit.

The other test involves setting max queue length to a value that the logging load would not hit.
At this higher max value (making sure it wont hit max for looping 100_000 logs, total of 190 MB), both `BlockingCollection` and `Queue` approaches get a big performance boost when logging the same amount of logs: 

```
Bytes written: 190 MB

Time elapsed using BlockingCollection (fastest after 10 reruns): 170 ms
Speed (fastest after 10 reruns): 1117.6471 MB/sec

Time elapsed using BlockingCollection (slowest after 10 reruns): 203 ms
Speed (slowest after 10 reruns): 935.96063 MB/sec

vs.

Time elapsed using Queue (fastest after 10 reruns): 169 ms
Speed (fastest after 10 reruns): 1124.2604 MB/sec

Time elapsed using Queue (slowest after 10 reruns): 201 ms
Speed (slowest after 10 reruns): 945.2736 MB/sec

vs.

System.Console
Time elapsed using System.Console: 4741 ms
Speed: 40.07593 MB/sec
```

Note: Given that these numbers are from manually using stopwatch, it doesn't cover the standard deviation error between using BlockingCollection and Queue. Overall both approaches remain in the same range in performance.

Fixes #65264

maryamariyan committed:
Enable configuring console logger's queue length and mode

    - Adds new APIs for the logging console queue processor
    - Replaces BlockingCollection with Queue implementation
    - Adds code coverage for ConsoleLoggerProcessor
    - Add queue name and metrics
    - Throw exception when new queue length is invalid

maryamariyan commented:
/cc @eerhardt @tarekgh @stephentoub @davidfowl @halter73

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerOptions.cs:
```diff
@@ -63,5 +63,29 @@ public ConsoleLoggerFormat Format
+        /// </summary>
+        public ConsoleLoggerBufferFullMode BufferFullMode { get; set; }
+
+        private int _maxQueuedMessages = 1048576;
```
>1048576 [](http://example.com/codeflow?start=41&length=7)

nit: write this as 1024 * 1024 and have a comment telling it is 1MB?

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerOptions.cs:
```diff
@@ -63,5 +63,29 @@ public ConsoleLoggerFormat Format
+            {
+                if (value <= 0)
+                {
+                    throw new ArgumentException(nameof(MaxQueueLength));
```
>MaxQueueLength [](http://example.com/codeflow?start=55&length=14)

do we use the property name? looking at other places in this file and seeing we are using `nameof(value)`

halter73 replied:
I think so. `ArgumentOutOfRangeException` is also probably slightly better.

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -12,16 +13,41 @@ namespace Microsoft.Extensions.Logging.Console
+        private readonly Queue<LogMessageEntry> _messageQueue;
+        private int _messagesDropped;
+        private bool _isAddingCompleted;
+        private int _maxQueuedMessages = 1048576;
```
>1048576 [](http://example.com/codeflow?start=41&length=7)

nit: please write in form can be easy to understand.

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -12,16 +13,41 @@ namespace Microsoft.Extensions.Logging.Console
+            {
+                if (value <= 0)
+                {
+                    throw new ArgumentException(nameof(MaxQueueLength));
```
>MaxQueueLength [](http://example.com/codeflow?start=55&length=14)

value?

martincostello replied:
If you used ArgumentOutOfRangeException you could use the constructor that accepts the value that was invalid, and also a message that explains what the valid range is.

stephentoub commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -12,16 +13,41 @@ namespace Microsoft.Extensions.Logging.Console
     internal class ConsoleLoggerProcessor : IDisposable
     {
-        private const int _maxQueuedMessages = 1024;
+        private static Meter _meter = new Meter("Microsoft-Extension-Logging-Console-Queue", "1.0.0");
```
```suggestion
        private static readonly Meter _meter = new Meter("Microsoft-Extension-Logging-Console-Queue", "1.0.0");
```

tarekgh replied:
>_meter [](http://example.com/codeflow?start=29&length=6)

nit: s_meter

stephentoub commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerOptions.cs:
```diff
@@ -63,5 +63,29 @@ public ConsoleLoggerFormat Format
+        /// <summary>
+        /// Gets or sets the desired console logger behavior when buffer becomes full. Defaults to <c>Wait</c>.
+        /// </summary>
+        public ConsoleLoggerBufferFullMode BufferFullMode { get; set; }
```
This should have validation in the setter that it's being set to either of the two supported values.

tarekgh commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -31,63 +57,133 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
+                var messagesDropped = Interlocked.Exchange(ref _messagesDropped, 0);
+                if (messagesDropped != 0)
+                {
+                    System.Console.Error.WriteLine($"{messagesDropped} message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this.{Environment.NewLine}");
```
>System.Console.Error.WriteLine($"{messagesDropped} message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this.{Environment.NewLine}"); [](http://example.com/codeflow?start=20&length=191)

Is it possible this message can spam the logs? I imagine showing this every time a message dropped can be a lot? is it right?

halter73 replied:
It's at most every other message, right? If I'm dropping logs between every message, I'd want that to be very clear. I don't like emitting this message any less.

maryamariyan replied:
it doesn't emit every time a message is dropped, but every other logs (could dropping 100, 900, etc. at a time)

halter73 replied:
Exactly!

It's also tempting to write this error exactly where the N logs were dropped. I could see that being *very* useful for debugging.

I understand we're trying to avoid adding more load to an already-overloaded system. In API review we discussed not wanting to add entries to the queue which might effectively halve the queue length. Still, I think it might be better to keep a reference to the tail `LogMessageEntry` in the `ConsoleLogProcessor` and store the `_messagesDropped` in that. `Enqueue` we could do something like:

```csharp
while (_messageQueue.Count >= MaxQueueLength && !_isAddingCompleted)
{
    if (FullMode == ConsoleLoggerBufferFullMode.DropWrite)
    {
        Interlocked.Increment(ref _tailLogMessageEntry._messagesDropped);
        return true;
    }
```

And then here we could do something like:

```csharp
if (entry._messagesDropped != 0)
{
    System.Console.Error.WriteLine($"{messagesDropped} message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this.");
}
```

I'm handwaving over how `_tailLogMessageEntry`, but this seems doable.

stephentoub commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -31,63 +57,133 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
+                if (messagesDropped != 0)
+                {
+                    System.Console.Error.WriteLine($"{messagesDropped} message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this.{Environment.NewLine}");
+                }
```
Is this really the right place for this?

maryamariyan replied:
Are you referring to whether or not `System.Console.Error` is the right place? I guess we could instead change it to write to `ErrorConsole` used in the line below it.

stephentoub commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -31,63 +57,133 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
+                        Interlocked.Increment(ref _messagesDropped);
+                        return true;
+                    }
+
```
Might want:
```C#
Debug.Assert(FullMode == ConsoleLoggerBufferFullMode.Wait);
```
here in case we add another mode in the future.

halter73 commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -12,16 +13,41 @@ namespace Microsoft.Extensions.Logging.Console
+                lock (_messageQueue)
+                {
+                    _maxQueuedMessages = value;
+                    Monitor.PulseAll(_messageQueue);
```
Is this meant to be reconfigurable at runtime? Should the `ConsoleLoggerProvider` be using `IOptionsMonitor` to call this setter when config changes?

maryamariyan replied:
Yes this is supposed to be reconfigurable.
`ConsoleLoggerProvider` uses `IOptionsMonitor`:

https://github.com/dotnet/runtime/blob/24f5de4bb10976dfb391e9b6affe5481f55a2d00/src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs#L41

In the reload callback we make sure to update the two properties in `ConsoleLoggerProcessor` accordingly.

TODO:

- [ ] Add test verifying it is reconfigurable.

halter73 commented on src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerExtensionsTests.cs:
```diff
@@ -356,6 +357,48 @@ public void AddConsole_NullFormatterNameUsingSystemdFormat_AnyDeprecatedProperti
+        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
+        [InlineData(-1)]
+        [InlineData(0)]
+        public void AddConsole_MaxQueueLengthSetToNegativeOrZero_IgnoresValue(int invalidMaxQueueLength)
```
```suggestion
        public void AddConsole_MaxQueueLengthSetToNegativeOrZero_Throws(int invalidMaxQueueLength)
```

It'd probably be easier to test this by calling `consoleLoggerOptions.MaxQueueLength = invalidMaxQueueLength;` directly, but I can appreciate verifying the end to end. We should probably still verify we're getting the right inner exception though.

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -31,63 +57,133 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
+            }
+            catch (Exception)
+            {
+                CompleteAdding();
```
If `console.Write(entry.Message);` on line 84 throws, then we call `CompleteAdding()` to gracefully stop logging. I believe that is also happening in the original code.

#### comparison with original code:

In the original code, `WriteMessage` is guarded by the try/catch block meant for any exception that might occur due to calling `GetConsumingEnumerable` (refer to [docs](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1.getconsumingenumerable?view=net-6.0)).

I believe if the `console.Write(entry.Message)` used to throw before the logger would gracefully stop due to this implementation. Even though it may not have been the intention then.

link to original code snippet:

https://github.com/dotnet/runtime/blob/24f5de4bb10976dfb391e9b6affe5481f55a2d00/src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs#L57-L80

tarekgh replied:
Looks we used to catch the exception inside ProcessLogQueue before and then CompleteAdding there. I guess the logic is moved now here instead. 

Maybe better to catch the exception in ProcessLogQueue as we did before? We didn't catch the exceptions in WriteMessage before

tarekgh replied:
I sent my comment without seeing @maryamariyan comment. 

> I think when console.Write(entry.Message) throws, in the original code, the logger gracefully stops due to this implementation. Even though it may not have been the intention then.

How does the logger stop at that time? could you please clarify with the following example? In the old code when calling `ConsoleLogger.Log` which will call `ConsoleLoggerProcessor.EnqueueMessage` which calls `WriteMessage`. If `WriteMessage` throws, nothing will happen and CompleteAdding will not be called.

maryamariyan replied:
The use case is described in unit the test called `ThrowDuringProcessLog_ShutsDownGracefully`

maryamariyan replied:
Just confirmed the behaviour in `ThrowDuringProcessLog_ShutsDownGracefully` is the same before or after the changes in this PR.

tarekgh replied:
I am not going to unresolve it but I hope I can get answer to the scenario I mentioned:

`In the old code when calling ConsoleLogger.Log which will call ConsoleLoggerProcessor.EnqueueMessage which calls WriteMessage. If WriteMessage throws, nothing will happen and CompleteAdding will not be called.` 

If the same exact scenario is executed with your change, I expect some different behavior as it will call `CompleteAdding`, I don't know if `ThrowDuringProcessLog_ShutsDownGracefully` is testing this exact scenario. I guess. Am I missing anything?

maryamariyan committed:
Apply PR feedback

maryamariyan commented:
Still a few more feedback left to apply. Marked resolved the ones that are done with the new commit.

maryamariyan committed:
More PR feedback

maryamariyan committed:
Code cleanup

maryamariyan committed:
Correct doc summary

maryamariyan committed:
bugfix

maryamariyan committed:
Write dropped notification to error console

noahfalk commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -31,63 +70,134 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
+                var messagesDropped = Interlocked.Exchange(ref _messagesDropped, 0);
+                if (messagesDropped != 0)
+                {
+                    System.Console.Error.WriteLine($"{messagesDropped} message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this.{Environment.NewLine}");
```
I think we should write this message to the IConsole and not to Console.Error. I also suggest that we write this message in exactly the position in the text where the other log messages have been omitted which would mean the warning message itself gets enqueued:

```C#
                 //in the Enqueue function
                if (_messageQueue.Count < MaxQueueLength)
                {
                    bool startedEmpty = _messageQueue.Count == 0;
                    if(_messagesDropped > 0)
                    {
                        _messageQueue.Enqueue(warningText);
                        _messagesDropped=0;
                    }
                    // if we just logged the dropped message warning this could push the queue size to 
                    // MaxLength+1, that shouldn't be a problem. It won't grow any further until it is less than
                    // MaxLength once again.
                    _messageQueue.Enqueue(item);
                    // if the queue started empty it could be at 1 or 2 now
                    if (startedEmpty)
                    {
                        // pulse for wait in Dequeue
                        Monitor.PulseAll(_messageQueue);
                    }

```

noahfalk replied:
I think the message should also mention that changing the BufferFullMode is an option.

noahfalk commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -31,63 +70,134 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
+                {
+                    if (FullMode == ConsoleLoggerBufferFullMode.DropWrite)
+                    {
+                        Interlocked.Increment(ref _messagesDropped);
```
If we change the dropped message handling via my suggestion above then we shouldn't need interlocked operations for this field any longer.

noahfalk commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs:
```diff
@@ -59,7 +55,15 @@ public ConsoleLoggerProvider(IOptionsMonitor<ConsoleLoggerOptions> options, IEnu
             }
-            _messageQueue = new ConsoleLoggerProcessor(console, errorConsole);
+            _messageQueue = new ConsoleLoggerProcessor(
+                "LoggingConsoleQueue",
```
If there is more than one processor I assume we want to give these unique names. Perhaps add an incrementing number on the end?

Logging, console, and queue are already in the names of the Meter and instrument so we could pick something more compact. How about "provider1", "provider2", etc?

maryamariyan replied:
There won't be more than one unless it's for testing. Because `ConsoleLoggerProcessor` is an internal class.

halter73 replied:
This is true if there's only one running host in the process which is true for the vast majority of apps using `Microsoft.Extensions.Logging.Console`. It is possible to run multiple hosts in a process though even in non-testing scenarios. I'm not sure how important we consider that to be.

noahfalk replied:
By more than one, I mean more than one in the entire process. Its true there will only be one Processor per ConsoleLoggerProvider, but ConsoleLoggerProvider is a public type and users can (and do) create multiple of these in the same process. As a test scenario create two LoggerFactories, add a new ConsoleLoggerProvider() to each one, and then log to both of them.

maryamariyan replied:
Fixed the code to properly support metrics when multiple console logger providers are present:

![image](https://user-images.githubusercontent.com/5897654/172692969-9cdb75b9-9691-4adf-b5c9-4bfdedae44af.png)

To get this I instantiated 3 different `LoggerFactory` instances each with their own console provider and processor as you described.

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerOptions.cs:
```diff
@@ -63,5 +63,42 @@ public ConsoleLoggerFormat Format
+            }
+        }
+
+        internal const int DefaultMaxQueueLengthValue = 1024 * 1024;
```
2500 is not enough for a default value. 

e.g. with

```c#
for (int i = 0; i < 10000; i++)
{
    logger.LogInformation(_1000Chars);
}
```

When testing to set the max size for the above `for` loop, the max size increase to `2500` won't be enough but for the same use case setting to 10,000 arbitrarily, makes sure we don't hit the lock contention and the speed jump that we want happens too.

maryamariyan replied:
I have no strong opinion of value default. Just sharing my observation here in case it helps decide on the best number to choose. Changed default from 1024*1024 to 2500 in latest commit.

cc @davidfowl

maryamariyan committed:
Apply PR feedback

- Changes default max queue length value
- Supports metrics for multiple console providers

maryamariyan commented:
Thanks for reviews. The PR is ready for another look.

noahfalk commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -12,16 +14,53 @@ namespace Microsoft.Extensions.Logging.Console
     internal class ConsoleLoggerProcessor : IDisposable
     {
-        private const int _maxQueuedMessages = 1024;
+        private static readonly Meter s_meter = new Meter("Microsoft-Extension-Logging-Console-Queue", "1.0.0");
```
This was marked resolved and you commented to take another look at the PR, but the code appears unchanged. Not sure if this was still a work in progress or it was missed.

noahfalk commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -31,63 +70,134 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
                 Name = "Console logger queue processing thread"
             };
             _outputThread.Start();
+            s_meter.CreateObservableGauge<long>("queue-size", GetQueueSize);
```
Same as above, this was marked resolved but still appears to be outstanding

maryamariyan replied:
This should be resolved now.

maryamariyan committed:
Apply remaining pr feedback

noahfalk commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -12,16 +14,56 @@ namespace Microsoft.Extensions.Logging.Console
-        private const int _maxQueuedMessages = 1024;
+        private static int s_processorCount;
+        private int _processorId;
+        private static List<ConsoleLoggerProcessor> s_processors = new List<ConsoleLoggerProcessor>();
```
This needs to be a list of weak references or something needs to remove Processors from the list at some point, otherwise the List could grow without bound. Silly code example:
```
for(int i = 0; i < 10_000_000_000; i++) { new ConsoleProvider(); } // this loop shouldn't OOM
```

noahfalk replied:
> Why is this static?

Because metrics are process global state.

maryamariyan replied:
removed s_processors

noahfalk commented on src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerProcessorTests.cs:
```diff
@@ -0,0 +1,176 @@
+            // Assert
+            if (okToDrop)
+            {
+                Assert.True(errorConsole.TimesWriteCalled > 0);
```
If the thread dequeing messages runs fast enough then nothing will be dropped and the test will fail. To make the test deterministic regardless of dequeue thread speed you would need to block at some point in the test console or console sink.

noahfalk commented:
Looking pretty good, just a few more things I spotted above.

noahfalk commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -31,63 +73,153 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
-                }
-                catch (InvalidOperationException) { }
-            }
+            s_meter.CreateObservableGauge<long>("queue-size", GetQueueSize);
```
I just double checked and I see we have several other counters in other areas of .NET that use "queue-length" but none which use "queue-size". I'd recommend changing this to match our precedent elsewhere.
```suggestion
            s_meter.CreateObservableGauge<long>("queue-length", GetQueueSize);
```

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -31,63 +73,153 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
         private void ProcessLogQueue()
         {
-            try
+            while (!_isAddingCompleted || _messageQueue.Count > 0)
```
Won't the thread complete if the `_messageQueue` is empty?

maryamariyan replied:
because that immediately takes us out before the first log message gets hit. that is not what we want

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -31,63 +73,153 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
 
-            // Adding is completed so just log the message
-            try
+        public virtual void EnqueueMessage(LogMessageEntry message)
```
Does this need to be `virtual`? Is it only for testing?

maryamariyan replied:
yes I believe it is used for testing

maryamariyan committed:
Remove static logic for metrics

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -35,59 +73,130 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
+
+        public bool Enqueue(LogMessageEntry item)
+        {
+            lock (_messageQueue)
+            {
+                while (_messageQueue.Count >= MaxQueueLength && !_isAddingCompleted)
+                {
+                    if (FullMode == ConsoleLoggerBufferFullMode.DropWrite)
+                    {
+                        _messagesDropped++;
+                        return true;
+                    }
+
+                    Debug.Assert(FullMode == ConsoleLoggerBufferFullMode.Wait);
+                    Monitor.Wait(_messageQueue);
```
The perf numbers between using custom queue implementation and blocking collection are shown to be the same.

one positive point for having the custom queue implementation is it makes it easier to allow configuration updates. I've discussed this with @eerhardt and @stephentoub and neither think that is necessarily a common feature to need to support.

If we think the custom queue implementation might pose a risk for future unknown race condition or bugs, I'm open to switching back to blocking collection.

noahfalk replied:
If we wanted to use BlockingCollection I expect we'd also have to rework enqueueing the warning message. Right now the lock is letting us atomically insert two messages into the queue + exceed the nominal limit by one message. I don't think it is that hard to make an alternate solution, but it does represent another unit of implementation work if we switched paths. I'm happy with whichever path you guys are more comfortable with.

maryamariyan replied:
> I don't think it is that hard to make an alternate solution, but it does represent another unit of implementation work if we switched paths.

I agree. I'm comfortable enough with the current state of the PR and did enough testing manual and unit tests.

I'll be open to revisiting implementation back to blocking collection in the future if need be.

maryamariyan reviewed (commented):
@noahfalk @davidfowl I'll switch the default to some number in the middle. I'd like this to be able to merge for the upcoming preview to give contributors a chance to try out and let us know how it fits in their solutions. that might influence a better default number prior to our final previews.

noahfalk commented:
> @noahfalk @davidfowl I'll switch the default to some number in the middle

Did I miss part of a conversation? I thought we had all agreed on 2500 and I hadn't seen any disagreement after that. Why the sudden large change to this value?

maryamariyan commented:
> Did I miss part of a conversation? I thought we had all agreed on 2500 and I hadn't seen any disagreement after that. Why the sudden large change to this value?

I personally don't think 2500 is that much different from 1024. they are both too low.

maryamariyan commented:
Pushing back to 2500 in the next commit

maryamariyan commented:
This PR is ready
- Default max queue size (decided 2500)
- Chosen implementation approach (custom queue).
- Adds support for updating full mode and queue size since the options are defined in `ConsoleLoggerOptions`.

noahfalk commented:
> Pushing back to 2500 in the next commit

👍 

> they are both too low

I think both you and David have felt this way, but I'm still not clear what leads you to this conclusion.

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/ref/Microsoft.Extensions.Logging.Console.cs:
```diff
@@ -62,6 +62,8 @@ public partial class ConsoleLoggerOptions
         public bool IncludeScopes { get { throw null; } set { } }
         public Microsoft.Extensions.Logging.LogLevel LogToStandardErrorThreshold { get { throw null; } set { } }
+        public int MaxQueueLength { get { throw null; } set { } }
+        public Microsoft.Extensions.Logging.Console.ConsoleLoggerBufferFullMode BufferFullMode { get { throw null; } set { } }
```
I believe the approved API has 

```C#
public ConsoleLoggerQueueFullMode QueueFullMode { get; set; }
```

This uses "queue" and "buffer". We should just have one name.

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerOptions.cs:
```diff
@@ -63,5 +63,42 @@ public ConsoleLoggerFormat Format
+            {
+                if (value != ConsoleLoggerBufferFullMode.Wait && value != ConsoleLoggerBufferFullMode.DropWrite)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(value), value, $"{nameof(value)} is not a supported buffer mode value.");
```
Can we put the exception messages in a .resx file?

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerTest.cs:
```diff
@@ -1107,26 +1107,6 @@ public void WriteCore_NullMessageWithNullException(LogLevel level)
         }
 
-        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
-        public void LogAfterDisposeWritesLog()
```
~Why's this test being deleted?~ Nevermind - it moved to ConsoleLoggerProcessorTests.cs.

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerProcessorTests.cs:
```diff
@@ -0,0 +1,176 @@
+using System.Linq;
+using Microsoft.Extensions.Logging.Test.Console;
+using Xunit;
+#pragma warning disable CS0618
```
Why are we disabling this warning for the whole file? Can it just be limited to where it is necessary?

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerProcessorTests.cs:
```diff
@@ -0,0 +1,176 @@
+            var console = new TestConsole(sink);
+            var errorConsole = new TimesWriteCalledConsole();
+            string queueName = nameof(CheckForNotificationWhenQueueIsFull) + (okToDrop ? "InDropWriteMode" : "InWaitMode");
+            var processor = new ConsoleLoggerProcessor(console, errorConsole, ConsoleLoggerBufferFullMode.Wait, 1024);
```
Why are we initializing the mode to `ConsoleLoggerBufferFullMode.Wait` here, but then changing it below on line 89? Why not just initialize it to the right value here?

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -12,16 +14,52 @@ namespace Microsoft.Extensions.Logging.Console
     internal class ConsoleLoggerProcessor : IDisposable
     {
-        private const int _maxQueuedMessages = 1024;
+        private const string WarningMessageOnDrop = " message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this. You may change `ConsoleLoggerBufferFullMode` to stop dropping messages.";
```
Should this go in a .resx file?

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -12,16 +14,52 @@ namespace Microsoft.Extensions.Logging.Console
-        private const int _maxQueuedMessages = 1024;
+        private const string WarningMessageOnDrop = " message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this. You may change `ConsoleLoggerBufferFullMode` to stop dropping messages.";
+        private readonly Queue<LogMessageEntry> _messageQueue;
+        private int _messagesDropped;
```
Does this need to be marked as `volatile` since it is read/write from multiple threads? Or since we only access it in a `lock`, are we OK?

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -12,16 +14,52 @@ namespace Microsoft.Extensions.Logging.Console
+                }
 
-        private readonly BlockingCollection<LogMessageEntry> _messageQueue = new BlockingCollection<LogMessageEntry>(_maxQueuedMessages);
+                _fullMode = value;
```
Does this need a `lock` and `Monitor.PulseAll` like `MaxQueueLength` has?

stephentoub replied:
In theory it should.  Otherwise, you could have a bunch of threads all blocked waiting for room to open up in the queue, and if the full mode is then changed to drop, all of those threads should wake up, throw their data away, and exit.

eerhardt reviewed (commented):
I just have some minor comments. Once those are addressed, I believe I will be able to approve.

It would also be good to get @stephentoub's sign off on the multi-threaded aspects of this in ConsoleLoggerProcessor.

maryamariyan committed:
Apply feedback - Rename to QueueFullMode

stephentoub commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -35,59 +73,130 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
-            catch (Exception) { }
         }
 
         // for testing
```
This method is only for testing, or it's internal rather than private for testing?  Seems like the latter, in which case it'd be good to change it to be:
```suggestion
        // internal for testing
```

stephentoub commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -35,59 +73,130 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
+                IConsole console = entry.LogAsError ? ErrorConsole : Console;
+                console.Write(entry.Message);
+            }
+            catch (Exception)
```
If we really want to eat everything, this can just be:
```suggestion
            catch
```

stephentoub commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -35,59 +73,130 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
         private void ProcessLogQueue()
         {
-            try
+            while (!_isAddingCompleted || _messageQueue.Count > 0)
```
I don't understand this code.  This seems both very thread-unsafe and unnecessary.  _messageQueue should only be accessed while holding the lock, and doesn't TryDequeue already do all the right checks?  Why isn't this just `while (TryDequeue(...))` rather than having TryDequeue be inside the if block?

maryamariyan committed:
Apply feedback - regex + pragma warning

stephentoub commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -35,59 +73,130 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
+                    Monitor.Wait(_messageQueue);
+                }
+
+                if (_messageQueue.Count < MaxQueueLength && !_isAddingCompleted)
+                {
```
Can this just be:
```suggestion
                if (!_isAddingCompleted)
                {
                    Debug.Assert(_messageQueue.Count < MaxQueueLength);
```
?

maryamariyan committed:
Apply PR feedback - correct test

maryamariyan committed:
Apply PR feedback - resx

maryamariyan committed:
Apply more feedback

maryamariyan committed:
Address thread safety issue

maryamariyan commented:
PR is ready for review again

noahfalk reviewed (approved):
LGTM to me in terms of diagnostic functionality : )

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/ref/Microsoft.Extensions.Logging.Console.cs:
```diff
@@ -77,6 +79,11 @@ public partial class ConsoleLoggerProvider : Microsoft.Extensions.Logging.ILogge
         public void Dispose() { }
         public void SetScopeProvider(Microsoft.Extensions.Logging.IExternalScopeProvider scopeProvider) { }
     }
+    public enum ConsoleLoggerBufferFullMode
```
```suggestion
    public enum ConsoleLoggerQueueFullMode
```

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/src/Resources/Strings.resx:
```diff
@@ -120,4 +120,10 @@
     <value>Cannot allocate a buffer of size {0}.</value>
   </data>
+  <data name="BufferModeNotSupported" xml:space="preserve">
+    <value>{0} is not a supported buffer mode value.</value>
```
```suggestion
    <value>{0} is not a supported queue mode value.</value>
```

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerOptions.cs:
```diff
@@ -63,5 +63,42 @@ public ConsoleLoggerFormat Format
+
+        private ConsoleLoggerBufferFullMode _bufferFullMode = ConsoleLoggerBufferFullMode.Wait;
+        /// <summary>
+        /// Gets or sets the desired console logger behavior when buffer becomes full. Defaults to <c>Wait</c>.
```
```suggestion
        /// Gets or sets the desired console logger behavior when the queue becomes full. Defaults to <c>Wait</c>.
```

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerOptions.cs:
```diff
@@ -63,5 +63,42 @@ public ConsoleLoggerFormat Format
+            {
+                if (value <= 0)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(value), value, $"{nameof(value)} must be larger than zero.");
```
Message should go into a resx

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerQueueFullMode.cs:
```diff
@@ -0,0 +1,20 @@
+        /// </summary>
+        Wait,
+        /// <summary>
+        /// Drops new log messages when the buffer is full
```
```suggestion
        /// Drops new log messages when the queue is full.
```

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/src/Resources/Strings.resx:
```diff
@@ -120,4 +120,10 @@
+    <value>{0} is not a supported buffer mode value.</value>
+  </data>
+  <data name="WarningMessageOnDrop" xml:space="preserve">
+    <value> message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this. You may change `ConsoleLoggerBufferFullMode` to stop dropping messages.</value>
```
```suggestion
    <value> message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this. You may change `ConsoleLoggerQueueFullMode` to stop dropping messages.</value>
```

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -12,16 +14,56 @@ namespace Microsoft.Extensions.Logging.Console
+                {
+                    // _fullMode is used inside the lock and is safer to guard setter with lock as well
+                    // this set is not expected to happen frequently
+                    _fullMode = value;
```
Why don't we need a `Monitor.PulseAll(_messageQueue);` here?

maryamariyan replied:
no thread would be waiting on full mode to change

maryamariyan replied:
oh seems like this comment is related, https://github.com/dotnet/runtime/pull/70186#discussion_r901990232
will update

maryamariyan committed:
rename

maryamariyan committed:
Apply suggestions from code review

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

maryamariyan committed:
Merge branch 'cloud-console' of https://github.com/maryamariyan/runtime into cloud-console

maryamariyan committed:
More feedback

maryamariyan committed:
feedback

eerhardt reviewed (approved):
LGTM. Thanks for the great work here, @maryamariyan!

JanKrivanek commented on src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs:
```diff
@@ -35,59 +78,128 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
+                    Monitor.Wait(_messageQueue);
+                }
+
+                if (_messageQueue.Count > 0 && !_isAddingCompleted)
```
So after the `Dispose` is called - we won't log any messages that are still in buffer?
@vlada-shubina as FYI

JanKrivanek replied:
To elaborate more
* `Dispose` calls `CompleteAdding` which tries to lock `_messageQueue` and then marks `_isAddingCompleted` with `true`
* `ProcessLogQueue` calls `TryDequeue` in a loop, per single item. For each call it needs to lock `_messageQueue` to be able to return item to be logged by `ProcessLogQueue`
* If `_isAddingCompleted` is `true` the `TryDequeue` will return false - even if `_messageQueue.Count` is nonempty

Those facts bring the following concerns:
* The `CompleteAdding` and `TryDequeue` races for same lock. Once `CompleteAdding` wins, nothing more is logged, even if items are still queued - this seems to be a bug.
* `TryDequeue` tries to acquire lock for each single item being logged - I'm surprised this ends up being faster then using low-lock `BlockingCollection`. If it really is - should the `BlockingCollection` be actualy revised?

tarekgh replied:
@JanKrivanek if you can provide a sample code to demonstrate the concerns, you may open a new issue and we can track looking at that there. Thanks for your feedback.

tarekgh replied:
I am seeing you already logged https://github.com/dotnet/runtime/issues/79812. We'll try to look at that at some point.

JanKrivanek replied:
Sample repro: https://github.com/JanKrivanek/ConsoleLogging

Showing that messages are being dropped in 7.0 version of logging.
